### ISSUE TITLE
Proposal: Implement CUDA Graphs for the PTX backend for zero-dispatch in `Taskgraphs` with multiple `Tasks`

### ISSUE THREAD

**Issue by mikepapadim** [beehive-lab/TornadoVM#780]
## Summary

Add support for [CUDA Graphs](https://docs.nvidia.com/cuda/cuda-programming-guide/04-special-topics/cuda-graphs.html) in TornadoVM's PTX backend to significantly reduce kernel launch overhead and improve performance for iterative workloads.

## Motivation

CUDA Graphs provide a mechanism to define a workflow as a graph of operations (kernel launches, memory copies, etc.) that can be instantiated once and launched repeatedly with minimal CPU overhead. This aligns naturally with TornadoVM's `TaskGraph` abstraction.

### Key Benefits

1. **Reduced Launch Overhead**: CUDA Graphs amortize the CPU-side setup costs across multiple executions. For graphs with many small kernels (common in ML inference, signal processing, and iterative solvers), this can dramatically reduce end-to-end latency.

2. **Whole-Workflow Optimization**: By presenting the entire workflow to CUDA at once, the driver can apply optimizations that aren't possible with piecewise stream submissions.

3. **Natural Fit with TaskGraph**: TornadoVM's `TaskGraph` already captures the dependency structure between tasks—this maps directly to CUDA Graph's node/edge model.

4. **Iterative Workloads**: Workloads like LLM inference (transformer layers executed repeatedly), physics simulations, and training loops benefit enormously from graph replay since the structure remains constant across iterations.

### Phase 1: Stream Capture Approach (Recommended Initial Implementation)

Use CUDA's stream capture mechanism to transparently build graphs from existing TaskGraph executions:

```
// Conceptual flow in PTX backend
cuStreamBeginCapture(stream, CU_STREAM_CAPTURE_MODE_GLOBAL);

// Execute TaskGraph operations (kernel launches, memcpy, etc.)
for (Task task : taskGraph.getTasks()) {
    launchKernel(task, stream);
}

cuStreamEndCapture(stream, &graph);
cuGraphInstantiate(&graphExec, graph, NULL, NULL, 0);

// Subsequent executions use the captured graph
cuGraphLaunch(graphExec, stream);
```

**Advantages**:
- Minimal changes to existing code generation
- Automatic capture of kernel dependencies
- Works with existing PTX kernel code

### Phase 2: Explicit Graph Construction (Future Enhancement)

For more control, implement explicit graph construction using the CUDA Graph API:

```
// Build graph nodes explicitly
cuGraphCreate(&graph, 0);
cuGraphAddKernelNode(&kernelNode, graph, dependencies, numDeps, &kernelParams);
cuGraphAddMemcpyNode(&memcpyNode, graph, dependencies, numDeps, &memcpyParams);
```

**Advantages**:
- Fine-grained control over node parameters
- Enables graph updates without re-instantiation
- Better suited for dynamic parameter updates

#### Option A: Automatic Graph Mode (Transparent)

Enable graph capture automatically when a TaskGraph is executed multiple times with the same structure:

```java
TaskGraph taskGraph = new TaskGraph("inference")
    .transferToDevice(DataTransferMode.FIRST_EXECUTION, input)
    .task("layer1", ComputeKernel::forward, input, weights1, output1)
    .task("layer2", ComputeKernel::forward, output1, weights2, output2)
    .transferToHost(DataTransferMode.EVERY_EXECUTION, output2);

ImmutableTaskGraph immutableGraph = taskGraph.snapshot();
TornadoExecutionPlan plan = new TornadoExecutionPlan(immutableGraph);

// First execution: captures the graph
plan.execute();

// Subsequent executions: replays the captured graph with minimal overhead
for (int i = 0; i < 1000; i++) {
    plan.execute();  // Uses CUDA Graph internally
}
```

#### Option B: Explicit Graph Mode (User Control)

Provide explicit API for users who want control over graph lifecycle:

```java
TornadoExecutionPlan plan = new TornadoExecutionPlan(immutableGraph)
    .withExecutionMode(ExecutionMode.CUDA_GRAPH);  // New execution mode

plan.captureGraph();  // Explicit capture
plan.executeGraph();  // Replay captured graph
plan.invalidateGraph();  // Force re-capture on next execution
```

### Implementation Components

1. **PTXStream Extension**
   - Add `beginCapture()` / `endCapture()` methods wrapping `cuStreamBeginCapture/cuStreamEndCapture`
   - Manage `CUgraph` and `CUgraphExec` handles

2. **PTXGraphCache**
   - Cache instantiated graphs keyed by TaskGraph signature
   - Handle invalidation when task parameters change significantly
   
3. **Graph Update Support**
   - Implement `cuGraphExecKernelNodeSetParams` for parameter updates without re-instantiation
   - Detect when topology changes require re-capture

4. **Memory Node Support**
   - Leverage CUDA Graph memory nodes for allocations within graphs
   - Handle `transferToDevice` / `transferToHost` as memcpy nodes

### Limitations and Considerations

Per CUDA documentation, graph nodes have constraints:

- **Kernel nodes**: Function context cannot change; no CUDA dynamic parallelism
- **Memcpy nodes**: Source/destination device cannot change between updates
- **Supported node types**: kernel, memcpy, memset, host function, empty, events, child graphs

TornadoVM TaskGraphs that only use kernel tasks and data transfers should be fully compatible.

### Testing Strategy

1. **Unit tests**: Verify graph capture/replay for single-kernel TaskGraphs
2. **Integration tests**: Multi-kernel graphs with dependencies
3. **Performance benchmarks**: 
   - Matrix multiplication with varying iteration counts
   - Transformer layer inference (multiple small kernels)
   - Compare stream-based vs graph-based execution

### Performance Expectations

Based on NVIDIA documentation and benchmarks:
- **Kernel launch overhead**: Reduced from ~5-10μs per kernel to <1μs amortized
- **Multi-kernel workloads**: Up to 10-50x reduction in CPU overhead for graphs with many small kernels
- **Memory bandwidth**: No change (GPU-side execution identical)

### 1. LLM Inference (GPULlama3.java and similar)

Transformer layers consist of many small operations (attention, FFN, normalization) executed in sequence. With CUDA Graphs:
- Capture the transformer layer computation once
- Replay for each token generation with updated KV-cache pointers

### 2. Iterative Solvers

Physics simulations, optimization algorithms, and numerical solvers that execute the same kernel sequence thousands of times.

### 3. Batch Processing

When processing multiple inputs with identical computation structure, a single graph can be replayed with updated input pointers.

## Related Work

- **JAX/XLA**: Uses CUDA Graphs for compiled function execution
- **PyTorch**: `torch.cuda.CUDAGraph` API for capturing and replaying CUDA operations  
- **TensorRT**: Uses CUDA Graphs internally for inference optimization

## Additional Resources

- [CUDA Graphs Documentation](https://docs.nvidia.com/cuda/cuda-programming-guide/04-special-topics/cuda-graphs.html)
- [CUDA Graphs Best Practices](https://developer.nvidia.com/blog/cuda-graphs/)
- [Stream Capture API Reference](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__GRAPH.html)
